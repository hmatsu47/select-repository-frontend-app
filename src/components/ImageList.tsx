import { For, Show } from "solid-js";
import Box from "@suid/material/Box";
import Button from "@suid/material/Button";
import Paper from "@suid/material/Paper";
import Table from "@suid/material/Table";
import TableBody from "@suid/material/TableBody";
import TableCell from "@suid/material/TableCell";
import TableContainer from "@suid/material/TableContainer";
import TableHead from "@suid/material/TableHead";
import TableRow from "@suid/material/TableRow";
import Typography from "@suid/material/Typography";
import {
  images,
  imageUri,
  isOpenedConfirm,
  repository,
  repositoryUri,
  setImageUri,
} from "../signal";
import { formatDateTimeDisplay } from "../formatDateTime";

export const ImageList = () => {
  return (
    <Show
      when={repository() && repositoryUri() && images() !== undefined}
      fallback={<></>}
    >
      <Box
        sx={{
          width: "100%",
          minWidth: "1024px",
          display: "flex",
        }}
        aria-live="polite"
      >
        <Show
          when={images() !== null}
          fallback={
            <Typography variant="h6">
              リポジトリにイメージがありません
            </Typography>
          }
        >
          <TableContainer component={Paper}>
            <Table aria-label="simple table">
              <TableHead>
                <TableRow>
                  <TableCell>選択</TableCell>
                  <TableCell>
                    イメージURI
                    <br />
                    （リポジトリURI:タグ または リポジトリURI@ダイジェスト）
                  </TableCell>
                  <TableCell>プッシュ日時</TableCell>
                  <TableCell>
                    サイズ
                    <br />
                    （MB）
                  </TableCell>
                  <TableCell>
                    脆弱性調査
                    <br />
                    （要AWSログイン）
                  </TableCell>
                </TableRow>
              </TableHead>
              <TableBody>
                <For each={images()} fallback={<></>}>
                  {(imageItem) => (
                    <TableRow>
                      <TableCell component="th" scope="row">
                        <Button
                          disabled={isOpenedConfirm()}
                          variant={
                            imageUri() && imageUri() === imageItem.uri
                              ? "contained"
                              : "outlined"
                          }
                          size="small"
                          color={
                            imageUri() && imageUri() === imageItem.uri
                              ? "primary"
                              : "inherit"
                          }
                          onClick={() => {
                            setImageUri(imageItem.uri);
                          }}
                          sx={{ textTransform: "none" }}
                          title={`${imageItem.uri} を選択`}
                        >
                          選択
                        </Button>
                      </TableCell>
                      <TableCell>{imageItem.uri}</TableCell>
                      <TableCell>
                        {formatDateTimeDisplay(new Date(imageItem.pushed_at))}
                      </TableCell>
                      <TableCell align="right">
                        {(
                          Math.round(
                            (imageItem.size / (1000 * 1000)) * Math.pow(10, 2)
                          ) / Math.pow(10, 2)
                        ).toLocaleString()}
                      </TableCell>
                      <TableCell>
                        <Button
                          disabled={isOpenedConfirm()}
                          variant="text"
                          onClick={() => {
                            const url = `https://${
                              repositoryUri()!.split(".")[3]
                            }.console.aws.amazon.com/ecr/repositories/private/${
                              repositoryUri()!.split(".")[0]
                            }/${repository()}/image/${
                              imageItem.digest
                            }/scan-results/?region=${
                              repositoryUri()!.split(".")[3]
                            }`;
                            window.open(url, "_blank");
                          }}
                          sx={{ textTransform: "none" }}
                          title={`${imageItem.digest} の結果を表示`}
                        >
                          結果表示
                        </Button>
                      </TableCell>
                    </TableRow>
                  )}
                </For>
              </TableBody>
            </Table>
          </TableContainer>
        </Show>
      </Box>
    </Show>
  );
};
