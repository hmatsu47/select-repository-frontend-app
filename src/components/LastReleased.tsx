import { Show } from "solid-js";
import Box from "@suid/material/Box";
import Stack from "@suid/material/Stack";
import Typography from "@suid/material/Typography";
import { formatDateTimeDisplay } from "../formatDateTime";
import { isReleased, lastImageUri, lastReleasedAt } from "../signal";

export const LastReleased = () => {
  return (
    <Show when={isReleased()} fallback={<></>}>
      <Box
        sx={{
          width: "100%",
          minWidth: "1024px",
          display: "flex",
        }}
        aria-live="polite"
      >
        <Stack spacing={2} direction="column">
          <Stack spacing={2} direction="row">
            <Typography variant="h6" sx={{ minWidth: "256px" }}>
              前回リリースイメージURI
            </Typography>
            <Typography variant="h6">{lastImageUri()!}</Typography>
          </Stack>
          <Stack spacing={2} direction="row">
            <Typography variant="h6" sx={{ minWidth: "256px" }}>
              前回リリース日時
            </Typography>
            <Typography variant="h6">
              {formatDateTimeDisplay(new Date(lastReleasedAt()!))}
            </Typography>
          </Stack>
        </Stack>
      </Box>
    </Show>
  );
};
