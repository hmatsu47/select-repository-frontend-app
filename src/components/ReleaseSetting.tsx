import { Show } from "solid-js";
import Box from "@suid/material/Box";
import Button from "@suid/material/Button";
import Stack from "@suid/material/Stack";
import Typography from "@suid/material/Typography";
import { formatDateTimeInput, formatDateTimeStore } from "../formatDateTime";
import {
  images,
  imageUri,
  isReleased,
  isReleaseSelected,
  releaseAt,
  setReleaseAt,
  setIsOpenedConfirm,
  setIsCancel,
} from "../signal";

export const ReleaseSetting = () => {
  const handleOpenSet = () => {
    setIsCancel(false);
    setIsOpenedConfirm(true);
  };
  const handleOpenCancel = () => {
    setIsCancel(true);
    setIsOpenedConfirm(true);
  };
  return (
    <Show when={isReleased() !== undefined && images()} fallback={<></>}>
      <Box
        sx={{
          width: "100%",
          minWidth: "1024px",
          display: "flex",
        }}
        aria-live="polite"
      >
        <Stack spacing={2} direction="column">
          <Stack
            spacing={2}
            direction="row"
            sx={{ display: "flex", alignItems: "center" }}
          >
            <Typography variant="h6" sx={{ minWidth: "256px", color: "#f00" }}>
              次回リリースイメージ URI
            </Typography>
            <Typography variant="subtitle1">
              {imageUri() ? imageUri() : "下のリストから選択してください"}
            </Typography>
          </Stack>
          <Stack spacing={2} direction="row">
            <Typography variant="h6" sx={{ minWidth: "256px", color: "#f00" }}>
              <label id="releaseAt">次回リリース日時</label>
            </Typography>
            <input
              type="datetime-local"
              aria-labelledby="releaseAt"
              value={
                releaseAt() ? formatDateTimeInput(new Date(releaseAt()!)) : ""
              }
              onChange={(event) => {
                setReleaseAt(
                  event.currentTarget.value === ""
                    ? undefined
                    : new Date(
                        Date.parse(
                          formatDateTimeStore(event.currentTarget.value)
                        )
                      )
                );
              }}
              max="9999-12-31 23:59"
            />
            <Button
              variant="contained"
              size="small"
              color="primary"
              onClick={() => {
                var tomorrow = new Date();
                tomorrow.setDate(tomorrow.getDate() + 1);
                tomorrow.setHours(4);
                tomorrow.setMinutes(5);
                tomorrow.setSeconds(0);
                tomorrow.setMilliseconds(0);
                setReleaseAt(tomorrow);
              }}
              sx={{ textTransform: "none" }}
              title="翌朝リリースする"
            >
              翌朝 04:05
            </Button>
            <Button
              variant="contained"
              size="small"
              color="secondary"
              onClick={() => {
                setReleaseAt(new Date());
              }}
              sx={{ textTransform: "none" }}
              title="即時リリースする"
            >
              現在（即時リリースする）
            </Button>
          </Stack>
          <Stack spacing={2} direction="row">
            <Typography variant="h6" sx={{ minWidth: "256px", color: "#f00" }}>
              次回リリースセット
            </Typography>
            <Button
              disabled={!imageUri() || !releaseAt()}
              variant="contained"
              size="small"
              color="primary"
              onClick={handleOpenSet}
              sx={{ textTransform: "none" }}
              title="次回リリースをセット"
            >
              指定のイメージ URI と日時をセット
            </Button>
            <Button
              disabled={!isReleaseSelected()}
              variant="contained"
              size="small"
              color="secondary"
              onClick={handleOpenCancel}
              sx={{ textTransform: "none" }}
              title="次回リリースを取り消し"
            >
              リリースを取り消し
            </Button>
          </Stack>
        </Stack>
      </Box>
    </Show>
  );
};
