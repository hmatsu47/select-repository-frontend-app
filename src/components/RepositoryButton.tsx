import Button from "@suid/material/Button";
import { fetchImages } from "../api/fetchImages";
import { repository, setRepository } from "../signal";
import { RepositoryItem } from "../type";

type Props = {
  repositoryItem: RepositoryItem;
};

export const RepositoryButton = (props: Props) => {
  return (
    <Button
      variant={
        repository() === props.repositoryItem.name ? "contained" : "outlined"
      }
      size="small"
      color={repository() === props.repositoryItem.name ? "primary" : "inherit"}
      onClick={async (e) => {
        setRepository(props.repositoryItem.name);
        await fetchImages();
      }}
      sx={{ textTransform: "none" }}
    >
      {props.repositoryItem.name}
    </Button>
  );
};
