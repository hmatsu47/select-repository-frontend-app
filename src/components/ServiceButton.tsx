import Button from "@suid/material/Button";
import { service, setService } from "../signal";
import { fetchRepositories } from "../api/fetchRepositories";
import { fetchSetting } from "../api/fetchSetting";
import { ServiceItem } from "../type";

type Props = {
  serviceItem: ServiceItem;
};

export const ServiceButton = (props: Props) => {
  return (
    <Button
      variant={service() === props.serviceItem.name ? "contained" : "outlined"}
      size="small"
      color="inherit"
      onClick={async (e) => {
        setService(props.serviceItem.name);
        await fetchSetting();
        await fetchRepositories();
      }}
      sx={{ textTransform: "none" }}
    >
      {props.serviceItem.name}
    </Button>
  );
};
