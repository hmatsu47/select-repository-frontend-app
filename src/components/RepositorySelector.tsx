import { For, Show } from "solid-js";
import Box from "@suid/material/Box";
import Stack from "@suid/material/Stack";
import Typography from "@suid/material/Typography";
import { RepositoryButton } from "./RepositoryButton";
import { repositories } from "../signal";

export const RepositorySelector = () => {
  return (
    <Box
      sx={{
        width: "100%",
        minWidth: "1024px",
        display: "flex",
      }}
      aria-live="polite"
    >
      <Show when={repositories()} fallback={<></>}>
        <Stack spacing={2} direction="row">
          <Typography variant="h6" sx={{ minWidth: "256px" }}>
            リリース対象リポジトリ
          </Typography>
          <For each={repositories()} fallback={<></>}>
            {(repositoryItem) => (
              <RepositoryButton repositoryItem={repositoryItem} />
            )}
          </For>
        </Stack>
      </Show>
    </Box>
  );
};
