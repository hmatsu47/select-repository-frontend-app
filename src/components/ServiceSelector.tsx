import { For, onMount, Show } from "solid-js";
import Box from "@suid/material/Box";
import Stack from "@suid/material/Stack";
import Typography from "@suid/material/Typography";
import { ServiceButton } from "./ServiceButton";
import { fetchServices } from "../api/fetchServices";
import { services, setService } from "../signal";

export const ServiceSelector = () => {
  onMount(async () => {
    await fetchServices();
    // ローカルストレージから前回の選択肢を読み取る
    const tmpService = localStorage.getItem("selectedService");
    if (tmpService) {
      setService(tmpService);
    }
  });

  return (
    <Box
      sx={{
        width: "100%",
        minWidth: "1024px",
        display: "flex",
      }}
      aria-live="polite"
    >
      <Show when={services()} fallback={<div>Loading...</div>}>
        <Stack spacing={2} direction="row">
          <Typography variant="h6" sx={{ minWidth: "256px" }}>
            リリース対象サービス
          </Typography>
          <For each={services()} fallback={<></>}>
            {(serviceItem) => <ServiceButton serviceItem={serviceItem} />}
          </For>
        </Stack>
      </Show>
    </Box>
  );
};
