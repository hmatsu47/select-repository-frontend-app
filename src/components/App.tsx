import { Component } from "solid-js";
import Box from "@suid/material/Box";
import Stack from "@suid/material/Stack";
import { ImageList } from "./ImageList";
import { LastReleased } from "./LastReleased";
import { Message } from "./Message";
import { ReleaseSetting } from "./ReleaseSetting";
import { RepositorySelector } from "./RepositorySelector";
import { ServiceSelector } from "./ServiceSelector";
import { TitleBar } from "./TitleBar";

export const App: Component = () => {
  return (
    <>
      <TitleBar />
      <Box
        sx={{
          padding: "10px 10px 0 10px",
          width: "100%",
          minWidth: "1024px",
          display: "flex",
          // justifyContent: "center",
        }}
        aria-live="polite"
      >
        <Stack spacing={2} direction="column">
          <ServiceSelector />
          <LastReleased />
          <RepositorySelector />
          <ReleaseSetting />
          <Message />
          <ImageList />
        </Stack>
      </Box>
    </>
  );
};
