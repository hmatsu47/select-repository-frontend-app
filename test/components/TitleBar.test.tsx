import { describe, expect, test } from "vitest";
import { render } from "@solidjs/testing-library";
import { formatSnapshot } from "../common/formatSnapshot";
import { TitleBar } from "../../src/components/TitleBar";

describe("<TitleBar />", () => {
  const titleBarList = [
    {
      title: "本番環境",
      environment: "本番環境",
      expected: "本番環境・コンテナリリース設定",
    },
    {
      title: "ステージング環境",
      environment: "ステージング環境",
      expected: "ステージング環境・コンテナリリース設定",
    },
    {
      title: "テスト環境",
      environment: "テスト環境",
      expected: "テスト環境・コンテナリリース設定",
    },
    {
      title: "指定なし",
      environment: undefined,
      expected: "コンテナリリース設定",
    },
  ];
  titleBarList.forEach((testCase) => {
    beforeEach(() => {
      // localStorage はモック化すべきかも（うまくいかなかったため一旦モックなしで実装）
      localStorage.removeItem("environment");
    });
    test(testCase.title, async () => {
      if (testCase.environment) {
        localStorage.setItem("environment", testCase.environment);
      }
      const { container, findByText, unmount } = render(() => <TitleBar />);
      const expected = (await findByText(testCase.expected)) as HTMLElement;
      expect(expected).toHaveTextContent(testCase.expected);
      // css の名前が動的に変わるので固定値に置換
      const html = formatSnapshot(container.innerHTML);
      expect(html).toMatchSnapshot();
      unmount();
    });
  });
  afterEach(() => {
    localStorage.removeItem("environment");
  });
});
