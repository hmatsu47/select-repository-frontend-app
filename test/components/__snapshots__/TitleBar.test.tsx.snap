// Vitest Snapshot v1

exports[`<TitleBar /> > ステージング環境 1`] = `"<header class=\\"MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic MuiAppBar-root MuiPaper-root css-xxxxxx\\"><div class=\\"MuiTypography-root MuiTypography-h6 MuiTypography-root css-xxxxxx\\">ステージング環境・コンテナリリース設定</div></header>"`;

exports[`<TitleBar /> > テスト環境 1`] = `"<header class=\\"MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic MuiAppBar-root MuiPaper-root css-xxxxxx\\"><div class=\\"MuiTypography-root MuiTypography-h6 MuiTypography-root css-xxxxxx\\">テスト環境・コンテナリリース設定</div></header>"`;

exports[`<TitleBar /> > 指定なし 1`] = `"<header class=\\"MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic MuiAppBar-root MuiPaper-root css-xxxxxx\\"><div class=\\"MuiTypography-root MuiTypography-h6 MuiTypography-root css-xxxxxx\\">コンテナリリース設定</div></header>"`;

exports[`<TitleBar /> > 本番環境 1`] = `"<header class=\\"MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic MuiAppBar-root MuiPaper-root css-xxxxxx\\"><div class=\\"MuiTypography-root MuiTypography-h6 MuiTypography-root css-xxxxxx\\">本番環境・コンテナリリース設定</div></header>"`;
